DM-29941: make the API RESTful

r-owen opened the pull request.

jonathansick commented on src/exposurelog/app.py:
```diff
@@ -3,17 +3,32 @@
-app.include_router(find_messages.router)
+subapp = fastapi.FastAPI(
+    title="Exposure log service",
+    description="A REST web service to create and manage log messages "
+    "that are associated with a particular exposure. "
+    "It is self-documenting using OpenAPI.",
```
```suggestion
    description="A REST web service to create and manage log messages "
    "that are associated with a particular exposure. ",
```

r-owen replied:
I would prefer that this match the README. So would prefer to change neither or both. See my comment above about your suggested change to the README.

jonathansick replied:
If anything, it'd be useful in this context for the root public GET endpoint to include in its body the URL to the OpenAPI docs.

r-owen replied:
Good suggestion. Done.

jonathansick commented on src/exposurelog/routers/delete_message.py:
```diff
@@ -11,33 +9,32 @@
-    Do this by setting ``is_valid`` False and updating
-    ``date_is_valid_changed``. Return the deleted messages.
+    Set ``is_valid`` False and update ``date_is_valid_changed``.
+    Return the deleted message.
```
The DELETE method doesn't return the deleted resource. Normally I'd recommend returning a 204 with an empty body. See https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE

r-owen replied:
Deleted messages are just marked as invalid. I thought it would be potentially useful to return the deleted resource to show that the is_valid flag was set False and the related datestamp was updated. Is it actively harmful to return this information?

r-owen replied:
Also DELETE is not quite idempotent -- the date_is_valid_set is always updated. I'm trying to fix that, but am not sure of how to do it with SQLAlchemy and still tell the difference between "was already deleted" and "record does not exist". In other words this is not good enough:
```
connection.execute(exposurelog_db.table
    .update()
    .where(sa.sql.and_(exposurelog_db.table.c.id == id, exposurelog_db.table.c.is_valid)
    .values(is_valid=False, date_is_valid_changed=current_time)
```

jonathansick replied:
> Deleted messages are just marked as invalid. I thought it would be useful to return the deleted resource to show that the is_valid flag was set False and the related timestamp was updated. Does that change your suggestion or would you still prefer I return 204 and an empty body?

It sounds to me like `is_valid` is more like an internal detail. I'd still just return `204` and an empty body.

r-owen replied:
OK. Done.

jonathansick reviewed (commented):
This is looking better; the focus on the `/messages` endpoint and using HTTP methods on that is definitely more RESTFul.

There's a couple of things that jump out to me and you can do with them as you feel appropriate.

First, I'm not sure about functionality of the `/messages/{site_id}_{id}` URL. Is there something that's relying on this particular ID format? And/or Is the idea that {id} isn't unique across sites? Personally, I'd either drop `{site_id}` from the URL and ensure that `{id}` is unique by making it a UUID rather than relying on the DB index (for instance), or if I do want `site_id` in the URL, I'd make it part of the endpoint structure, like `/sites/{site_id}/messages/{id}`. Again, if there's an external requirement that IDs be `{site_id}_{id}` that's fine.

Second, I notice that this API is using form encodings everywhere rather than JSON? Is this a requirement? For an actual RESTFul API I'd expect JSON rather than form encodings that make this service more a backend for a very specific website. Is this a requirement from the LOVE interface? Given that they're probably writing LOVE in some sort of JS framework, I'd think it'd be easy (and probably preferable) for them to second JSON requests rather than a form.

timj commented:
Re UUID, butler registry is migrating completely to UUID primary keys.

r-owen commented on README.rst:
```diff
@@ -2,8 +2,8 @@
-Exposure log is a web service to create and manage log messages associated with a particular exposure.
-It is self-documented using OpenAPI.
+Exposure log is a REST web service to create and manage log messages that are associated with a particular exposure.
+It is self-documenting using OpenAPI.
```
How would a user know to look for the docs unless this information is in README?

jonathansick replied:
I recommended removing it because it's a superfluous fact about the app that doesn't add to the description of what it does (you could also write that it's built with FastAPI and Python and deployed with Kubernetes... but none of those facts help to describe the app as well as the first sentence did. 

The very next paragraph of the README provides a path for the HTTP API documentation, and that's really you need to get the point across.

r-owen replied:
Good point. Fixed.

r-owen commented:
I didn't know about UUIDs. That seems like a nicer approach than making the primary key out of an auto-incrementing integer plus the site ID. It also dramatically reduces the cost of setting the site ID wrong (i.e. duplicating the same site ID at two different locations). It will require a new schema, but it's a small change.

timj commented:
lsst/daf_butler#500 was where we changed butler over to UUID. It has some snippets in there for making sqlalchemy use UUID (which is native on postgres but a string in sqlalchey).

r-owen commented:
Thank you for catching the Form issue; I didn't realize there was a difference and I certainly want json. I switched all such parameters to Body and updated the two affected unit tests accordingly.

I will switch to UUID.

r-owen commented:
I have updated the code to use UUID. I have left the field name as "id" but would be delighted to pick something else, such as "uuid" if that is more common or more idiomatic. (It would be a small relief to avoid the conflict with the rarely-used built in `id` function).

jonathansick reviewed (approved):
Looks good, I like the refactoring of the prefix in the router too.

r-owen commented:
A URL to docs is a bit iffy because OpenAPI apparently has several different flavors of docs at different URLs. I suppose I could pick one. In any case I'll add OpenAPI to the data returned by the main link.

r-owen commented:
OK. I think i made all suggested changes.

The root URL now serves HTML that includes a link to one of the OpenAPI docs (which looks much nicer than the dict it used to return and should be fine until we have a need to return something else). I removed the OpenAPI reference in the README (because the OpenAPI doc link is given a few lines down) and the app metadata (no need for it there, since the root now returns the info).

DELETE returns 204 and is (finally) idempotent. I also made `is_valid` computed so I only ever set `date_invalidated` (which I renamed from `date_is_valid_changed`; it starts out NULL and will only ever be set once).

Thank you very much for the thorough review and helpful suggestions.
